clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1004,43 +1126,53 @@ static bool buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
-          Context.getDiagnostics().Report(Fld->getLocation(),
-                                          diag::err_sycl_non_std_layout_type)
-              << ArgTy;
-          AllArgsAreValid = false;
-          continue;
-        }
-      }
-
-      CXXRecordDecl *RD =
-          cast<CXXRecordDecl>(ArgTy->getAs<RecordType>()->getDecl());
-      if (!RD->hasTrivialCopyConstructor()) {
-        Context.getDiagnostics().Report(
-            Fld->getLocation(),
-            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
-            << 0 << ArgTy;
+      if (!copyableToKernel(Fld, ArgTy)) {
         AllArgsAreValid = false;
         continue;
       }
-      if (!RD->hasTrivialDestructor()) {
-        Context.getDiagnostics().Report(
-            Fld->getLocation(),
-            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
-            << 1 << ArgTy;
-        AllArgsAreValid = false;
-        continue;
-      }
-
       CreateAndAddPrmDsc(Fld, ArgTy);
-
       // Create descriptors for each accessor field in the class or struct
       createParamDescForWrappedAccessors(Fld, ArgTy);
     } else if (ArgTy->isReferenceType()) {
-      Context.getDiagnostics().Report(
-          Fld->getLocation(), diag::err_bad_kernel_param_type) << ArgTy;
+      Context.getDiagnostics().Report(Fld->getLocation(),
+                                      diag::err_bad_kernel_param_type)
+          << ArgTy;
       AllArgsAreValid = false;
+    } else if (ArgTy->isConstantArrayType()) {
+      const ConstantArrayType *CAT = cast<ConstantArrayType>(ArgTy);
+      QualType ET = CAT->getElementType();
+      if (Util::isSyclAccessorType(ET)) {
+        size_t NumElements = static_cast<size_t>(CAT->getSize().getZExtValue());
+        // Treat each element of accessor array as a separate item
+        // Create a parameter descriptor for it
+        for (size_t I = 0; I < NumElements; ++I) {
+          createSpecialSYCLObjParamDesc(Fld, ET);
+        }
+      } else {
+        if (!copyableToKernel(Fld, ET)) {
+          AllArgsAreValid = false;
+          continue;
+        }
+        auto wrapAnArray = [&](const QualType ArgTy, RecordDecl *&NewClass,
+                               FieldDecl *&Field) {
+          NewClass = Context.buildImplicitRecord("wrapped_array");
+          NewClass->startDefinition();
+          Field = FieldDecl::Create(
+              Context, NewClass, SourceLocation(), SourceLocation(),
+              /*Id=*/nullptr, ArgTy,
+              Context.getTrivialTypeSourceInfo(ArgTy, SourceLocation()),
+              /*BW=*/nullptr, /*Mutable=*/false, /*InitStyle=*/ICIS_NoInit);
+          Field->setAccess(AS_public);
+          NewClass->addDecl(Field);
+          NewClass->completeDefinition();
+        };
+        RecordDecl *NewClass;
+        FieldDecl *Field;
+        wrapAnArray(ArgTy, NewClass, Field);
+        QualType FT = Context.getRecordType(NewClass);
+        wrapAnArray(FT, NewClass, Field);
```

Comment:
The doc says "array is actually double-wrapped in structs. This was done simply to preserve the interface to an existing function within the code which processes each kernel caller parameter as a capture object member."
I still don't fully understand what it means.
Could you please elaborate a bit more? Which function do you mean here?

Reply:
This is the function being used:
CreateAndAddPrmDsc = [&](const FieldDecl *Fld, const QualType &ArgType)

The object being added to a list inside that function is Fld, which is expected to be a Field of some struct. So a wrapped struct cannot be passed to this function. A wrapped wrapped struct is needed. This does not affect the generated code.
struct {
  struct {
    int array[100];
  }
}
As I said, this could be changed by me, but it would mean changes to the CreateAndAddPrmDsc implementation, to all callers of CreateAndAddPrmDsc, and to the place where the list created by this function is processed.

Reply:
Oh, right. I understand, thanks. I think we need some refactoring for this all, but I won't insist on doing this along with your change. Could you please add a comment to the code describes why arrays are double wrapped?

Reply:
Comments have been added in SemaSYCL.cpp explaining double-wrapped of arrays.

Reply:
Thanks!